crates/compression/src/tables.rs

```diff
@@ -0,0 +1,210 @@
+use fuel_core_types::{
+    fuel_compression::{
+        CompressibleBy,
+        DecompressibleBy,
+        RegistryKey,
+    },
+    fuel_tx::{
+        input::PredicateCode,
+        Address,
+        AssetId,
+        ContractId,
+        ScriptCode,
+    },
+};
+use std::collections::HashMap;
+
+use crate::{
+    compress::{
+        CompressCtx,
+        CompressDb,
+        PrepareCtx,
+    },
+    decompress::{
+        DecompressCtx,
+        DecompressDb,
+        DecompressError,
+    },
+    ports::{
+        EvictorDb,
+        TemporalRegistry,
+    },
+};
+
+macro_rules! tables {
+    ($($name:ident: $type:ty),*$(,)?) => {
+        #[doc = "RegistryKey namespaces"]
+        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, serde::Serialize, serde::Deserialize)]
+        #[allow(non_camel_case_types)] // Match names in structs exactly
+        pub enum RegistryKeyspace {
+            $(
+                $name,
+            )*
+        }
+
+        #[doc = "RegistryKey namespace with an associated typed value"]
+        #[derive(Debug, Clone, PartialEq, Eq, Hash, serde::Serialize, serde::Deserialize)]
+        #[allow(non_camel_case_types)] // Match names in structs exactly
+        pub enum RegistryKeyspaceValue {
+            $(
+                $name($type),
+            )*
+        }
+
+        #[doc = "A value for each keyspace"]
+        #[derive(Debug, Clone, Default)]
+        pub struct PerRegistryKeyspace<T> {
+            $(pub $name: T,)*
+        }
+        impl<T> core::ops::Index<RegistryKeyspace> for PerRegistryKeyspace<T> {
+            type Output = T;
+
+            fn index(&self, index: RegistryKeyspace) -> &Self::Output {
+                match index {
+                    $(
+                        RegistryKeyspace::$name => &self.$name,
+                    )*
+                }
+            }
+        }
+        impl<T> core::ops::IndexMut<RegistryKeyspace> for PerRegistryKeyspace<T> {
+            fn index_mut(&mut self, index: RegistryKeyspace) -> &mut Self::Output {
+                match index {
+                    $(
+                        RegistryKeyspace::$name => &mut self.$name,
+                    )*
+                }
+            }
+        }
+
+        #[doc = "Key-value mapping for each keyspace"]
+        #[derive(Debug, Clone, Default)]
+        pub struct PerRegistryKeyspaceMap {
+            $(pub $name: HashMap<RegistryKey, $type>,)*
+        }
+
+        impl PerRegistryKeyspaceMap {
+            #[cfg(test)]
+            pub fn insert(&mut self, key: RegistryKey, value: RegistryKeyspaceValue) {
```

Comment:
Where do we call this function?

Reply:
https://github.com/FuelLabs/fuel-core/blob/0957584e90b3ed344cd5e2e31502ffc13958b127/crates/compression/src/lib.rs#L133